app/src/main/java/uk/gov/onelogin/login/ui/welcome/WelcomeScreenViewModel.kt

```diff
@@ -102,38 +103,73 @@ class WelcomeScreenViewModel @Inject constructor(
-                else -> {
-                    _loading.emit(false)
-                    loginSession.present(
-                        launcher,
-                        configuration = LoginSessionConfiguration(
-                            authorizeEndpoint = authorizeEndpoint,
-                            clientId = clientId,
-                            locale = locale,
-                            redirectUri = redirectUri,
-                            scopes = scopes,
-                            tokenEndpoint = tokenEndpoint,
-                            persistentSessionId = persistentId
-                        )
+            handleGetClientAttestation {
+                loginSession.present(
+                    launcher,
+                    configuration = LoginSessionConfiguration(
+                        authorizeEndpoint = authorizeEndpoint,
+                        clientId = clientId,
+                        locale = locale,
+                        redirectUri = redirectUri,
+                        scopes = scopes,
+                        tokenEndpoint = tokenEndpoint,
+                        persistentSessionId = persistentId
                     )
-                }
+                )
             }
         }
     }
 
-    @Suppress("TooGenericExceptionCaught")
     fun handleActivityResult(intent: Intent, isReAuth: Boolean = false) {
         if (intent.data == null) return
 
         viewModelScope.launch {
+            val savedAttestation = appIntegrity.retrieveSavedClientAttestation()
+            // Attempt to get a new attestation if the saved one is not available due to device or open secure store
+            // Very unlikely to occur
+            if (savedAttestation.isNullOrEmpty()) {
+                handleGetClientAttestation { attestation ->
+                    handleCreatePoP(intent, attestation, isReAuth)
+                }
+                // Attestation retrieved successfully, directly create PoP
+            } else {
+                handleCreatePoP(intent, savedAttestation, isReAuth)
+            }
+        }
+    }
+
+    private suspend fun handleGetClientAttestation(
+        onSuccess: (String) -> Unit
+    ) {
+        when (val attestation = appIntegrity.getClientAttestation()) {
+            is AttestationResult.Failure -> {
+                _loading.emit(false)
+                navigator.navigate(LoginRoutes.SignInError)
+            }
+            is AttestationResult.NotRequired -> {
+                _loading.emit(false)
+                onSuccess(
+                    attestation.savedAttestation
+                )
+            }
+            else -> {
+                _loading.emit(false)
+                onSuccess(
+                    (attestation as AttestationResult.Success).clientAttestation
+                )
+            }
+        }
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private fun handleCreatePoP(intent: Intent, attestation: String, isReAuth: Boolean) {
```

Comment:
I think this method needs to be refactored, it kind of breaks the single responsibility principle here. I think this method needs to work as above and take a callback method